takari-builder/src/main/java/io/takari/builder/internal/maven/MavenDependencyResolver.java

```diff
@@ -0,0 +1,308 @@
+    } catch (DependencyResolutionException e) {
+      result = e.getResult();
+    }
+
+    Set<Artifact> artifacts = new LinkedHashSet<>();
+    if (result.getDependencyGraph() != null
+        && !result.getDependencyGraph().getChildren().isEmpty()) {
+      RepositoryUtils.toArtifacts(artifacts, result.getDependencyGraph().getChildren(),
+          Collections.singletonList(project.getArtifact().getId()), resolutionFilter);
+    }
+    return artifacts;
+  }
+
+  private DependencyFilter getResolutionFilter(ResolutionScope scope, boolean transitive) {
+    DependencyFilter filter = new ScopeDependencyFilter(null, negate(scope));
+    if (!transitive) {
+      filter = AndDependencyFilter.newInstance(filter,
+          new TransitiveDependencyFilter(project.getDependencies()));
+    }
+
+    return filter;
+  }
+
+  private static Collection<String> negate(ResolutionScope resolutionScope) {
+    Collection<String> result = new HashSet<>();
+    Collections.addAll(result, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_COMPILE,
+        Artifact.SCOPE_PROVIDED, Artifact.SCOPE_RUNTIME, Artifact.SCOPE_TEST);
+
+    for (String scope : toScopes(resolutionScope)) {
+      if (Artifact.SCOPE_COMPILE.equals(scope)) {
+        result.remove(Artifact.SCOPE_COMPILE);
+        result.remove(Artifact.SCOPE_SYSTEM);
+        result.remove(Artifact.SCOPE_PROVIDED);
+      } else if (Artifact.SCOPE_RUNTIME.equals(scope)) {
+        result.remove(Artifact.SCOPE_COMPILE);
+        result.remove(Artifact.SCOPE_RUNTIME);
+      } else if (Artifact.SCOPE_TEST.equals(scope)) {
+        result.clear();
+      }
+    }
+
+    return result;
+  }
+
+  public static Collection<String> toScopes(ResolutionScope scope) {
+    if (ResolutionScope.COMPILE.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_PROVIDED);
+    } else if (ResolutionScope.RUNTIME.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
+    } else if (ResolutionScope.COMPILE_PLUS_RUNTIME.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_PROVIDED,
+          Artifact.SCOPE_RUNTIME);
+    } else if (ResolutionScope.TEST.equals(scope)) {
+      return Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_SYSTEM, Artifact.SCOPE_PROVIDED,
+          Artifact.SCOPE_RUNTIME, Artifact.SCOPE_TEST);
+    }
+    return Collections.emptyList();
+  }
+
+  private DependencyResolutionResult resolveDependencies(ResolutionScope scope,
+      DependencyFilter resolutionFilter) throws DependencyResolutionException {
+    ClassLoader tccl = Thread.currentThread().getContextClassLoader();
+    try {
+      ClassLoader projectRealm = project.getClassRealm();
+      if (projectRealm != null && projectRealm != tccl) {
+        Thread.currentThread().setContextClassLoader(projectRealm);
+      }
+
+      RequestTrace trace = RequestTrace.newChild(null, resolutionFilter);
```

Comment:
I would pass in a supplier rather than have a hard code construction mehtod:
of RequestTrace.newChild()  ->   traceSupplier.supply(null, resolutionFilter);